rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/CostAccounting.scala

```diff
@@ -1,28 +1,46 @@
 package coop.rchain.rholang.interpreter.accounting
 
-import cats.effect.Sync
-import cats.effect.concurrent.Ref
+import cats.data._
+import cats.effect.Concurrent
+import cats.effect.concurrent._
 import cats.implicits._
 import cats.mtl._
 import cats.{FlatMap, Monad}
 import coop.rchain.rholang.interpreter.errors.OutOfPhlogistonsError
 
 object CostAccounting {
 
-  def of[F[_]: Sync](init: Cost): F[MonadState[F, Cost]] =
+  private[this] def of[F[_]: Concurrent](init: Cost): F[MonadState[F, Cost]] =
     Ref[F]
       .of(init)
-      .map(
-        state =>
-          new DefaultMonadState[F, Cost] {
-            val monad: cats.Monad[F]                      = implicitly[Monad[F]]
-            def get: F[Cost]                              = state.get
-            def set(s: Cost): F[Unit]                     = state.set(s)
-            override def modify(f: Cost => Cost): F[Unit] = state.update(f)
-          }
-      )
-
-  def empty[F[_]: Sync]: F[MonadState[F, Cost]] =
-    this.of(Cost(0, "init"))
+      .map(defaultMonadState)
 
+  private[this] def empty[F[_]: Concurrent]: F[MonadState[F, Cost]] =
+    Ref[F]
+      .of(Cost(0, "init"))
+      .map(defaultMonadState)
+
+  def emptyCost[F[_]: Concurrent](implicit L: FunctorTell[F, Chain[Cost]]): F[_cost[F]] =
+    for {
+      s <- Semaphore(1)
+      c <- empty
+    } yield (loggingCost(c, L, s))
+
+  def initialCost[F[_]: Concurrent](
+      init: Cost
+  )(implicit L: FunctorTell[F, Chain[Cost]]): F[_cost[F]] =
+    for {
+      s <- Semaphore(1)
```

Comment:
why `Fall back to Ref, use Semaphore for synchronization`? I assume Sync is still needed when we do the fall back? If so - I'm not sure I see the benefit?

Reply:
Sync provides the `bracketCase` operation

Reply:
This doesn't answer why a fallback to Ref + Semaphore is needed / better tham MVar. I don't have preference, I want to understand (and then I might have preference)

Reply:
It's described in the "caveats" section of this PR